packages/arb-token-bridge-ui/src/components/TransferPanel/TransferPanel.tsx

```diff
@@ -372,6 +389,213 @@ export function TransferPanel() {
+    if (!selectedToken) {
+      return
+    }
+    if (!l1Signer || !l2Signer) {
+      throw 'Signer is undefined'
+    }
+
+    const isDeposit = type === 'deposits'
+
+    setTransferring(true)
+    let currentNetwork = isDeposit
+      ? latestNetworksAndSigners.current.l1.network
+      : latestNetworksAndSigners.current.l2.network
+
+    const currentNetworkName = getNetworkName(currentNetwork.id)
+    const isConnectedOnTheWrongChain =
+      (isDeposit && isConnectedToArbitrum.current) ||
+      (type === 'withdrawals' && !isConnectedToArbitrum.current)
+
+    if (isConnectedOnTheWrongChain) {
+      if (shouldTrackAnalytics(currentNetworkName)) {
+        trackEvent('Switch Network and Transfer', {
+          type: isDeposit ? 'Deposit' : 'Withdrawal',
+          tokenSymbol: 'USDC',
+          assetType: 'ERC-20',
+          accountType: isSmartContractWallet ? 'Smart Contract' : 'EOA',
+          network: currentNetworkName,
+          amount: Number(amount)
+        })
+      }
+      const switchTargetChainId = isDeposit
+        ? latestNetworksAndSigners.current.l1.network.id
+        : latestNetworksAndSigners.current.l2.network.id
+      try {
+        await switchNetworkAsync?.(switchTargetChainId)
+        currentNetwork = isDeposit
+          ? latestNetworksAndSigners.current.l1.network
+          : latestNetworksAndSigners.current.l2.network
+      } catch (e) {
+        if (isUserRejectedError(e)) {
+          return
+        }
+      }
+    }
+
+    try {
+      const l1ChainID = latestNetworksAndSigners.current.l1.network.id
+      const l2ChainID = latestNetworksAndSigners.current.l2.network.id
+      const sourceChainId = isDeposit ? l1ChainID : l2ChainID
+
+      const waitForInput = isDeposit
+        ? openUSDCDepositConfirmationDialog()
+        : openUSDCWithdrawalConfirmationDialog()
+      const [confirmed, primaryButtonClicked] = await waitForInput()
+
+      if (!confirmed) {
+        return
+      }
+      if (isDeposit && primaryButtonClicked === 'bridged') {
+        // User has selected normal bridge (USDC.e)
+        depositToken()
+        return
+      }
+
+      // CCTP has an upper limit for transfer
+      const burnLimit = await fetchPerMessageBurnLimit({
+        sourceChainId
+      })
+
+      if (burnLimit.lte(amountBigNumber)) {
+        const formatedLimit = formatAmount(burnLimit, {
+          decimals: selectedToken.decimals,
+          symbol: 'USDC'
+        })
+        errorToast(
+          `The limit for transfers using CCTP is ${formatedLimit}. Please lower your amount and try again.`
+        )
+        return
+      }
+
+      const recipient = destinationAddress || walletAddress
+      const { usdcContractAddress, tokenMessengerContractAddress } =
+        getContracts(sourceChainId)
+
+      const allowance = await getTokenAllowanceForSpender({
+        account: walletAddress,
+        erc20Address: usdcContractAddress,
+        provider: latestConnectedProvider.current,
```

Comment:
Since we know the direction of the transfer, why not just use the provider based on that? This way we're using the provider of the currently connected network, and that may cause some weird edge cases

Reply:
Good point